clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -2622,36 +2649,113 @@ raw2trace_t::thread_file_at_eof(raw2trace_thread_data_t *tdata)
+                while (erase_from > 0 &&
+                       tdata->delayed_branch[erase_from - 1].type ==
+                           TRACE_TYPE_ENCODING) {
+                    log(4, "Erasing cached encoding for %p\n",
+                        tdata->delayed_branch_decode_pcs[instr_index]);
+                    tdata->encoding_emitted.erase(
+                        tdata->delayed_branch_decode_pcs[instr_index]);
+                    --erase_from;
+                }
+                VPRINT(
+                    4,
+                    "Discarded %zd entries for final branch without subsequent instr\n",
+                    tdata->delayed_branch.size() - erase_from);
+                tdata->delayed_branch.erase(tdata->delayed_branch.begin() + erase_from,
+                                            tdata->delayed_branch.end());
+                tdata->delayed_branch_decode_pcs.pop_back();
+                tdata->delayed_branch_target_pcs.pop_back();
+                break;
             } else {
+                if (target == nullptr) {
+                    DEBUG_ASSERT(!type_is_instr_direct_branch(
+                        static_cast<trace_type_t>(entry.type)));
+                    trace_entry_t local[3];
+                    int size = trace_metadata_writer_t::write_marker(
+                        reinterpret_cast<byte *>(local), TRACE_MARKER_TYPE_BRANCH_TARGET,
+                        reinterpret_cast<uintptr_t>(next_instr_pc));
+                    DEBUG_ASSERT(static_cast<size_t>(size) <= sizeof(local));
+                    for (int local_idx = 0;
+                         local_idx < size / static_cast<int>(sizeof(local[0]));
+                         ++local_idx) {
+                        tdata->delayed_branch.insert(tdata->delayed_branch.begin() + i,
+                                                     local[local_idx]);
+                    }
+                    VPRINT(4, "Inserted indirect branch target %p\n", next_instr_pc);
+                } else if (entry.type == TRACE_TYPE_INSTR_CONDITIONAL_JUMP) {
+                    if (target == next_instr_pc) {
+                        branch_type = TRACE_TYPE_INSTR_TAKEN_JUMP;
+                    } else {
+                        branch_type = TRACE_TYPE_INSTR_UNTAKEN_JUMP;
+                    }
+                    entry.type = static_cast<unsigned short>(branch_type);
+                }
                 VPRINT(4,
-                       "Appending delayed branch tagalong entry type %s (%d) for thread "
-                       "%d\n",
-                       trace_type_names[entry.type], entry.type, tdata->index);
+                       "Appending delayed branch type=%d pc=%p decode=%p target=%p for "
+                       "thread %d\n",
+                       branch_type, branch_addr,
+                       tdata->delayed_branch_decode_pcs[instr_index], target,
+                       tdata->index);
             }
+            next_instr_pc = branch_addr;
```

Comment:
This assumes that all delayed instrs are branches, which is true, but let's add the above ASSERT(false) in else.

Reply:
Where does it assume they are all branches?  Agreed that they are all branches, but I think it would work if a non-branch were stuck in there (if it had its fall-through PC stored for its _target_pcs).  It would be surprising though.  Added a DEBUG_ASSERT that it's a branch up above inside the `if (type_is_instr`.